linera-service/src/linera-exporter/state.rs

```diff
@@ -81,4 +147,161 @@ where
+        match result {
+            None => Ok(None),
+            Some(Value::Block(block)) => Ok(Some(block)),
+            _ => Err(ViewError::InconsistentEntries),
+        }
+    }
+
+    pub async fn get_blob(&mut self, blob: BlobId) -> Result<Option<&mut Blob>, ViewError> {
+        let result = self.index.get_mut(&blob.into()).await?;
+        match result {
+            None => Ok(None),
+            Some(Value::Blob(blob)) => Ok(Some(blob)),
+            _ => Err(ViewError::InconsistentEntries),
+        }
+    }
+
+    pub async fn contains_block(&self, hash: CryptoHash) -> Result<bool, ViewError> {
+        match self.index.get(&hash.into()).await? {
+            None => Ok(false),
+            Some(Value::Block(_)) => Ok(true),
+            _ => Err(ViewError::InconsistentEntries),
+        }
+    }
+
+    pub async fn contains_blob(&self, blob: BlobId) -> Result<bool, ViewError> {
+        match self.index.get(&blob.into()).await? {
+            None => Ok(false),
+            Some(Value::Blob(_)) => Ok(true),
+            _ => Err(ViewError::InconsistentEntries),
+        }
+    }
+}
+
+impl From<Block> for Value {
+    fn from(value: Block) -> Self {
+        Self::Block(value)
+    }
+}
+
+impl From<Blob> for Value {
+    fn from(value: Blob) -> Self {
+        Self::Blob(value)
+    }
+}
+
+impl From<CryptoHash> for Key {
+    fn from(value: CryptoHash) -> Self {
+        Key::Block(value)
+    }
+}
+
+impl From<BlobId> for Key {
+    fn from(value: BlobId) -> Self {
+        Key::Blob(value)
+    }
+}
+
+pub(super) struct Batch {
+    block: ConfirmedBlock,
+    block_dependencies: Vec<ConfirmedBlock>,
+    blob_dependencies: Vec<ConfirmedBlob>,
+}
+
+impl Batch {
+    pub(super) fn new(block: ConfirmedBlock) -> Batch {
+        Self {
+            block,
+            block_dependencies: Vec::new(),
+            blob_dependencies: Vec::new(),
+        }
+    }
+
+    pub(super) fn add_block(&mut self, block: ConfirmedBlock) {
+        self.block_dependencies.push(block);
+    }
+
+    pub(super) fn add_blob(&mut self, blob: ConfirmedBlob) {
+        self.blob_dependencies.push(blob);
+    }
+
+    #[cfg(with_testing)]
+    pub(super) fn into_summary(self) -> Summary {
+        let mut summary = Summary::new(&self.block);
+
+        for block in self.block_dependencies {
+            summary.add_block(&block);
+        }
+
+        for blob in self.blob_dependencies {
+            summary.add_blob(&blob);
+        }
+
+        summary
+    }
+}
+
+#[async_trait]
+pub(super) trait StorageExt {
+    type Key;
+    type Batch;
+    type Error;
+
+    async fn read_batch(&self, batch: Vec<Self::Key>) -> Result<Self::Batch, Self::Error>;
+}
+
+#[async_trait]
+impl<S> StorageExt for S
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    type Key = Key;
+    type Batch = Batch;
+    type Error = ExporterError;
+
+    async fn read_batch(&self, keys: Vec<Self::Key>) -> Result<Self::Batch, Self::Error> {
```

Comment:
seems like you may want to use `read_confirmed_blocks` and `read_blobs` (plural) here